fs: improve errors thrown from fs.watch()

joyeecheung opened the pull request:
- Add an accessor property `initialized `to FSEventWrap to
  check the state of the handle from the JS land
- Introduce ERR_FS_WATCHER_ALREADY_STARTED so calling start()
  on a watcher that is already started will throw instead of
  doing nothing silently.
- Introduce ERR_FS_WATCHER_NOT_STARTED so calling close()
  on a watcher that is already closed will throw instead of
  doing nothing silently.
- Validate the filename passed to fs.watch()
- Assert that the handle in the watcher are instances of
  FSEvent instead of relying on the illegal invocation error
  from the VM.
- Add more assertions in FSEventWrap methods now that we check
  `initialized` and the filename in JS land before invoking
  the binding.
- Use uvException instead of errornoException to create
  the errors with the error numbers from libuv to make them
  consistent with other errors in fs.

TODO:

- Improve fs.watchFile() the same way this patch improves fs.watch()
- It seems possible to fire both rename and change event from libuv
  together now that we can check if the handle is closed via
  `initialized` in JS land.

This is a retake of https://github.com/nodejs/node/pull/17851

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
fs

joyeecheung commented on lib/fs.js:
```diff
@@ -1305,11 +1311,17 @@ function FSWatcher() {
   this._handle.owner = this;
 
   this._handle.onchange = function(status, eventType, filename) {
+    // TODO(joyeecheung): we may check self._handle.initialized here
```
cc @bnoordhuis can you check that this can be fixed this way? Thanks!

joyeecheung replied:
BTW, this is coming from:
https://github.com/nodejs/node/blob/9e9c516512668133813c71e39d56b011b804572b/src/fs_event_wrap.cc#L161-L171

joyeecheung commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/13450/

bnoordhuis commented on lib/fs.js:
```diff
@@ -1328,21 +1340,34 @@ FSWatcher.prototype.start = function(filename,
-                               persistent,
-                               recursive,
+                               !!persistent,
+                               !!recursive,
```
Minor if correct change in behavior.

Right now, C++ land calls IsTrue(), which is essentially a strict `x === true` check.  Using coercion here changes the observable behavior to a non-strict `x == true` check.

joyeecheung replied:
@bnoordhuis Yes, generally I think for boolean options it seems to be more intuitive to coerce them as if they are handled like `if (options.something)`. The default cases are set in `fs.watch` so we should be fine in the common cases.

BridgeAR replied:
I personally think it is best to have a strict check. Why would we want to accept any thruthy value?

bnoordhuis commented on src/fs_event_wrap.cc:
```diff
@@ -95,6 +105,18 @@ void FSEventWrap::Initialize(Local<Object> target,
+      FIXED_ONE_BYTE_STRING(env->isolate(), "initialized"),
+      get_initialized_templ,
+      Local<FunctionTemplate>(),
+      static_cast<PropertyAttribute>(ReadOnly | DontDelete | v8::DontEnum));
```
Is there a reason to make it an accessor instead of a method?

joyeecheung replied:
@bnoordhuis No strong reasons other than this way you can't assign it to something else or wrap it...probably plays more nicely with inspector?

joyeecheung commented:
@nodejs/tsc Needs one more TSC approval to land since it's semver-major..thanks!

mcollina commented on src/fs_event_wrap.cc:
```diff
@@ -105,22 +127,19 @@ void FSEventWrap::New(const FunctionCallbackInfo<Value>& args) {
 }
 
-
+// wrap.start(filename, persistent, recursive, encoding)
```
this line is commented out. Can you remove it?

joyeecheung replied:
@mcollina It's the signature of this method...probably less obvious than I thought

BridgeAR replied:
I think it is nice to have the signature. I am not sure what the convention for this is in C++ though. Something similar to JSDoc might be nicer?

joyeecheung replied:
This is more about the positions, the types are pretty obvious given the assertions below.

BridgeAR replied:
I do not have a strong opinion... hm... maybe change `wrap.start` to: `Arguments: `?

joyeecheung commented:
CITGM against master: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1323/
CITGM against this PR: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1324/

BridgeAR commented on test/parallel/test-fs-watch-enoent.js:
```diff
@@ -1,21 +1,56 @@
+    assert.strictEqual(err.code, 'ENOENT');
+    assert.strictEqual(err.syscall, 'watch');
+    return true;
+  };
```
As a suggestion: `assert.throws` supports the object notation. So instead of using the function here, it could also be written as:

```js
assert.throws(
  () => assert.fail(),
  {
    path: 'foo',
    filename: 'bar',
    ...
  }
)
```

joyeecheung replied:
With functions it will be easier to adjust the validations to different platforms, e.g. some throws ENOENT while some throws ENODEV (AIX)

BridgeAR replied:
I am not sure in what way that is easier to validate but it is only a suggestion, so *shrug*.

joyeecheung replied:
@BridgeAR  It can be written like this

```
if (err.code === 'ENOENT') {
   // validate ENOENT
} else {
  // validate ENODEV
}
```

BridgeAR commented on lib/fs.js:
```diff
@@ -1328,21 +1340,34 @@ FSWatcher.prototype.start = function(filename,
                                      persistent,
                                      recursive,
                                      encoding) {
+  lazyAssert()(this._handle instanceof FSEvent, 'handle must be a FSEvent');
```
Since very recently we have the `(D)CHECK` macro. I think it would be nice to replace these assert calls with those instead :-).

joyeecheung replied:
@BridgeAR Not unless `CHECK` throws instead of aborting...at least currently we only expects an error thrown in `sequential/test-fs-watch.js`

BridgeAR replied:
Well... it is a semver-major PR. But this is also just a suggestion.

BridgeAR commented on test/sequential/test-fs-watch.js:
```diff
@@ -112,12 +112,15 @@ tmpdir.refresh();
 {
   let oldhandle;
-  assert.throws(function() {
+  common.expectsError(() => {
```
This change is actually not necessary. `assert.throws` does accept objects as well.

joyeecheung commented:
Reverted the non-strict check and added `ENODEV` checks to work around the AIX bug.

New CI: https://ci.nodejs.org/job/node-test-pull-request/13549/

BridgeAR commented on test/parallel/test-fs-watch-enoent.js:
```diff
@@ -1,21 +1,64 @@
+
+  const validateError = (err) => {
+    assert.strictEqual(nonexistentFile, err.path);
+    assert.strictEqual(nonexistentFile, err.filename);
```
Nit: the arguments should be the other way around. Otherwise the error message would be weird in case of an error.

richardlau commented on test/parallel/test-fs-watch-enoent.js:
```diff
@@ -1,21 +1,64 @@
+        err.message,
+        `ENODEV: no such device, watch '${nonexistentFile}'`);
+      assert.strictEqual(err.errno, uv.UV_ENODEV);
+      assert.strictEqual(err.code, 'ENODEV');
```
@nodejs/platform-aix is this expected behaviour?

gireeshpunathil replied:
```patch
--- a/test/parallel/test-fs-watch-enoent.js
+++ b/test/parallel/test-fs-watch-enoent.js
@@ -6,6 +6,7 @@ const fs = require('fs');
 assert.throws(function() {
   fs.watch('non-existent-file');
 }, function(err) {
+  console.log(err)
   assert(err);
   assert(/non-existent-file/.test(err));
   assert.strictEqual(err.filename, 'non-existent-file');
```

with a console.log in the callback of the existing code, this is what I get in AIX:

```trace
{ Error: watch non-existent-file ENODEV
    at FSWatcher.start (fs.js:1346:19)
    at Object.fs.watch (fs.js:1369:11)
    at /tmp/gireesh/node/test/parallel/test-fs-watch-enoent.js:7:6
    at getActual (assert.js:421:5)
    at Function.throws (assert.js:430:18)
    at Object.<anonymous> (/tmp/gireesh/node/test/parallel/test-fs-watch-enoent.js:6:8)
    at Module._compile (module.js:670:30)
    at Object.Module._extensions..js (module.js:681:10)
    at Module.load (module.js:581:32)
    at tryModuleLoad (module.js:521:12)
  errno: 'ENODEV',
  code: 'ENODEV',
  syscall: 'watch non-existent-file',
  filename: 'non-existent-file' }
```
hope this helps.

gireeshpunathil replied:
I just tested with the changes from this PR on a local AIX box, and this test looks good - the error messages are matching, thanks!

joyeecheung commented:
Only an infra issue on ubuntu1604-32 but it was green before the update. Just in case, another Linux CI: https://ci.nodejs.org/job/node-test-commit-linux/16930/

joyeecheung commented:
There seem to be no new CITGM errors. Linux CI is clean, landing..

joyeecheung commented:
Landed in 6c25f2ea49c2521dfd2423bf3a06222633ec4dc9, thanks!
